python/lsst/pipe/base/execution_reports.py

```diff
@@ -0,0 +1,434 @@
+        --------
+        TaskExecutionReport.inspect_quantum()
+        """
+        status_graph.nodes[output_ref.id]["not_produced"] = False
+        self.n_produced += 1
+
+
+@dataclasses.dataclass
+class TaskExecutionReport:
+    """A report on the status and content of a task in an executed quantum
+    graph.
+
+    Use task metadata to identify and inspect failures and report on output
+    datasets.
+
+    See Also
+    --------
+    QuantumGraphExecutionReport
+    DatasetTypeExecutionReport
+    """
+
+    failed: dict[uuid.UUID, DatasetRef] = dataclasses.field(default_factory=dict)
+    """A mapping from quantum data ID to log dataset reference for quanta that
+    failed directly in this run (`dict`).
+    """
+    failed_upstream: dict[uuid.UUID, DataCoordinate] = dataclasses.field(default_factory=dict)
+    """A mapping of data IDs of quanta that were not attempted due to an
+    upstream failure (`dict`).
+    """
+    output_datasets: dict[str, DatasetTypeExecutionReport] = dataclasses.field(default_factory=dict)
+    """Missing and produced outputs of each DatasetType (`dict`).
+    """
+
+    def inspect_quantum(
+        self,
+        quantum_node: QuantumNode,
+        status_graph: networkx.DiGraph,
+        refs: Iterable[DatasetRef],
+        metadata_name: str,
+        log_name: str,
+    ) -> None:
+        """Inspect a quantum of a quantum graph and ascertain the status of
+        each associated data product.
+
+        Parameters
+        ----------
+        quantum_node: `QuantumNode`
+            The specific node of the quantum graph to be inspected.
+        status_graph: `networkx.DiGraph`
+            The quantum graph produced by
+            QuantumGraphExecutionReport.make_reports() which steps through the
+            quantum graph of a run and logs the status of each quantum.
+        refs: `Iterable[DatasetRef]`
+            The DatasetRefs of each of the DatasetTypes produced by the task.
+            Includes initialization, intermediate and output data products.
+        metadata_name: `str`
+            The metadataDataset name for the node.
+        log_name: `str`
+            The name of the log files for the node.
+
+        See Also
+        --------
+        DatasetTypeExecutionReport.handle_missing_dataset()
+        DatasetTypeExecutionReport.handle_produced_dataset()
+        QuantumGraphExecutionReport.make_reports()
+        """
+        quantum = quantum_node.quantum
+        (metadata_ref,) = quantum.outputs[metadata_name]
+        (log_ref,) = quantum.outputs[log_name]
+        if metadata_ref.id not in refs[metadata_name]:
+            if any(
+                status_graph.nodes[upstream_quantum_id]["failed"]
+                for upstream_dataset_id in status_graph.predecessors(quantum_node.nodeId)
+                for upstream_quantum_id in status_graph.predecessors(upstream_dataset_id)
+            ):
+                self.failed_upstream[quantum_node.nodeId] = quantum.dataId
```

Comment:
The mypy warning here is telling you that `Quantum` does not require a dataId be defined.  That means that this can return `None` but you have declared that `failed_upstream` can not hold `None` but must always be `DataCoordinate`.

I think to appease mypy you will need to do something like

```python
if quantum.dataId is not None:
    self.failed_upstream[...] = ...
```
or else (because you know this must be impossible)

```python
assert quantum.dataId is not None
```
(you know this must be impossible since you are getting the quantum from a graph but mypy can't know that).